File: ZadorozhnyyAlexander/Block 3/Task 3/filters.c
#pragma once
#include "structs.h"
#include "image_functions.h"
#include "filters.h"
#include <math.h>

int compare(const void* a, const void* b) {
	return *(int*)a - *(int*)b;
}

unsigned char greyscale(rgb_colors rgb)
{
	return ((0.3 * rgb.blue) + (0.6 * rgb.green) + (0.1 * rgb.red));
}

void grey_filter(image picture)
{
	for (int i = 0; i < picture.height; i++)
	{
		for (int j = 0; j < picture.width; j++)
		{
			picture.rgb[i][j].red = picture.rgb[i][j].green = picture.rgb[i][j].blue = greyscale(picture.rgb[i][j]);
		}
	}
}

image sobel_filter(image picture, int direction)
{
	grey_filter(picture);
	image new_pic = copy_image(picture);

	int edge = 80;
	//int g_x[3][3] = { {1, 0, -1}, {2, 0, -2}, {1, 0, -1} };
	//int g_y[3][3] = { {1, 2, 1}, {0, 0, 0}, {-1, -2, -1} };
	for (int i = 1; i < picture.height - 1; i++)
	{
		for (int j = 1; j < picture.width - 1; j++)
		{
			/*int sum_x = 0, sum_y = 0;
			for (int k = -1; k <= 1; k++)
			{
				for (int p = -1; p <= 1; p++)
				{
					sum_x += picture.rgb[i + k][j + p].red * g_x[k + 1][p + 1];
					sum_y += picture.rgb[i + k][j + p].red * g_y[k + 1][p + 1];
				}
			}*/
			int sum_x = (picture.rgb[i + 1][j - 1].red + 2 * picture.rgb[i + 1][j].red + picture.rgb[i + 1][j + 1].red) -
				(picture.rgb[i - 1][j - 1].red + 2 * picture.rgb[i - 1][j].red + picture.rgb[i - 1][j + 1].red);
			int sum_y = (picture.rgb[i - 1][j + 1].red + 2 * picture.rgb[i][j + 1].red + picture.rgb[i + 1][j + 1].red) -
				(picture.rgb[i - 1][j - 1].red + 2 * picture.rgb[i][j - 1].red + picture.rgb[i + 1][j - 1].red);
			switch (direction)
			{
			case -1:
				new_pic.rgb[i][j].red = new_pic.rgb[i][j].green = new_pic.rgb[i][j].blue = abs(sum_x) <= edge ? 0 : 255;
				break;
			case 0:
				new_pic.rgb[i][j].red = new_pic.rgb[i][j].green = new_pic.rgb[i][j].blue = sqrt(pow(sum_x, 2) + pow(sum_y, 2)) <= edge ? 0 : 255;
				break;
			case 1:
				new_pic.rgb[i][j].red = new_pic.rgb[i][j].green = new_pic.rgb[i][j].blue = abs(sum_y) <= edge ? 0 : 255;
				break;
			}
		}
	}
	free_image(picture);
	return new_pic;
}

void middle_filter(image picture)
{
	for (int i = 1; i < picture.height - 1; i++)
	{
		for (int j = 1; j < picture.width - 1; j++)
		{
			int mid_arr[3][9];
			int index = 0;
			for (int k = -1; k <= 1; k++)
			{
				for (int p = -1; p <= 1; p++)
				{
					mid_arr[0][index] = picture.rgb[i + k][j + p].red;
					mid_arr[1][index] = picture.rgb[i + k][j + p].green;
					mid_arr[2][index] = picture.rgb[i + k][j + p].blue;
					index++;
				}
			}
			qsort(mid_arr[0], 9, sizeof(int), compare);
			qsort(mid_arr[1], 9, sizeof(int), compare);
			qsort(mid_arr[2], 9, sizeof(int), compare);
			picture.rgb[i][j].red = mid_arr[0][4];
			picture.rgb[i][j].green = mid_arr[1][4];
			picture.rgb[i][j].blue = mid_arr[2][4];
		}
	}
}

void gausses_filter_3x3(image picture)
{
	int g[3][3] = { {1, 2, 1}, {2, 4, 2}, {1, 2, 1} };
	for (int i = 1; i < picture.height - 1; i++)
	{
		for (int j = 1; j < picture.width - 1; j++)
		{
			int r_k = 0, g_k = 0, b_k = 0;
			for (int k = -1; k <= 1; k++)
			{
				for (int p = -1; p <= 1; p++)
				{
					r_k += picture.rgb[i + k][j + p].red * g[k + 1][p + 1];
					g_k += picture.rgb[i + k][j + p].green * g[k + 1][p + 1];
					b_k += picture.rgb[i + k][j + p].blue * g[k + 1][p + 1];
				}
			}
			picture.rgb[i][j].red = r_k / 16;
			picture.rgb[i][j].green = g_k / 16;
			picture.rgb[i][j].blue = b_k / 16;
		}
	}
}

File: ZadorozhnyyAlexander/Semester 2/Block 1/Filters/BMPFileFilter/Filters.cs
using System;

namespace BMPFileFilter
{
    public class Filters
    {
        private static byte GreyScale(Rgb pixel)
        {
            return Convert.ToByte((0.3 * pixel.Blue) + (0.6 * pixel.Green) + (0.1 * pixel.Red));
        }

        public static void ApplyGreyFilter(BitMapFile file)
        {
            for (int i = 0; i < file.Height; i++)
            {
                for (int j = 0; j < file.CountRowRgb; j++)
                {
                    file.ImageRgb[i, j].Red = file.ImageRgb[i, j].Green = file.ImageRgb[i, j].Blue = GreyScale(file.ImageRgb[i, j]);
                }
            }
        }

        private static void CopyList(Rgb[,] first, Rgb[,] second)
        {
            for (int i = 0; i < first.GetLength(0); i++)
            {
                for (int j = 0; j < first.GetLength(1); j++)
                {
                    second[i, j] = new Rgb(first[i, j].GetListBytes());
                }
            }
        }

        public static void ApplySobelFilter(BitMapFile file, Direction type)
        {
            ApplyGreyFilter(file);
            Rgb[,] newPic = new Rgb[file.ImageRgb.GetLength(0), file.ImageRgb.GetLength(1)];
            CopyList(file.ImageRgb, newPic);

            int edge = 80;
            for (int i = 1; i < file.Height - 1; i++)
            {
                for (int j = 1; j < file.CountRowRgb - 1; j++)
                {
                    int sumX = (file.ImageRgb[i + 1, j - 1].Red + 2 * file.ImageRgb[i + 1, j].Red + file.ImageRgb[i + 1, j + 1].Red) -
                        (file.ImageRgb[i - 1, j - 1].Red + 2 * file.ImageRgb[i - 1, j].Red + file.ImageRgb[i - 1, j + 1].Red);
                    int sumY = (file.ImageRgb[i - 1, j + 1].Red + 2 * file.ImageRgb[i, j + 1].Red + file.ImageRgb[i + 1, j + 1].Red) -
                        (file.ImageRgb[i - 1, j - 1].Red + 2 * file.ImageRgb[i, j - 1].Red + file.ImageRgb[i + 1, j - 1].Red);

                    switch ((int)type)
                    {
                        case 0:
                            newPic[i, j].Red = newPic[i, j].Green = newPic[i, j].Blue = Math.Abs(sumX) <= edge ? (byte)0 : (byte)255;
                            break;
                        case 1:
                            newPic[i, j].Red = newPic[i, j].Green = newPic[i, j].Blue = Math.Sqrt(Math.Pow(sumX, 2) + Math.Pow(sumY, 2)) <= edge ? (byte)0 : (byte)255;
                            break;
                        case 2:
                            newPic[i, j].Red = newPic[i, j].Green = newPic[i, j].Blue = Math.Abs(sumY) <= edge ? (byte)0 : (byte)255;
                            break;
                    }
                }
            }
            CopyList(newPic, file.ImageRgb);
        }

        public static void ApplyMiddleFilter(BitMapFile file)
        {
            for (int i = 1; i < file.Height - 1; i++)
            {
                for (int j = 1; j < file.CountRowRgb - 1; j++)
                {
                    byte[] midArrRed = new byte[9], midArrGreen = new byte[9], midArrBlue = new byte[9];
                    int index = 0;
                    for (int k = -1; k <= 1; k++)
                    {
                        for (int p = -1; p <= 1; p++)
                        {
                            midArrRed[index] = file.ImageRgb[i + k, j + p].Red;
                            midArrGreen[index] = file.ImageRgb[i + k, j + p].Green;
                            midArrBlue[index] = file.ImageRgb[i + k, j + p].Blue;
                            index++;
                        }
                    }

                    Array.Sort(midArrRed);
                    Array.Sort(midArrGreen);
                    Array.Sort(midArrBlue);

                    file.ImageRgb[i, j].Red = midArrRed[4];
                    file.ImageRgb[i, j].Green = midArrGreen[4];
                    file.ImageRgb[i, j].Blue = midArrBlue[4];
                }
            }
        }

        public static void ApplyGauss3x3Filter(BitMapFile file)
        {
            int[,] gaussMatrix = { { 1, 2, 1}, { 2, 4, 2}, { 1, 2, 1} };
            for (int i = 1; i < file.Height - 1; i++)
            {
                for (int j = 1; j < file.CountRowRgb - 1; j++)
                {
                    int redKoef = 0, greenKoef = 0, blueKoef = 0;
                    for (int k = -1; k <= 1; k++)
                    {
                        for (int p = -1; p <= 1; p++)
                        {
                            redKoef += file.ImageRgb[i + k, j + p].Red * gaussMatrix[k + 1, p + 1];
                            greenKoef += file.ImageRgb[i + k, j + p].Green * gaussMatrix[k + 1, p + 1];
                            blueKoef += file.ImageRgb[i + k, j + p].Blue * gaussMatrix[k + 1, p + 1];
                        }
                    }
                    file.ImageRgb[i, j].Red = Convert.ToByte(redKoef / 16);
                    file.ImageRgb[i, j].Green = Convert.ToByte(greenKoef / 16);
                    file.ImageRgb[i, j].Blue = Convert.ToByte(blueKoef / 16);
                }
            }
        }

        public enum Direction
        {
            X,
            Both,
            Y
        }
    }
}
